### PR TITLE
Enable '--strictNullChecks'

### LINKED ISSUE #9432: Turn on --strictNullChecks in codebase
This is important for consumers of our API and for testing the feature in general

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #9432

The changes generally amount to:

* Adding `| undefined` as appropriate.
* Adding `!` in many places
	- Particularly on optional members of interfaces like `Symbol` and `Type`.
	- Also, flags enums are often used as `E | undefined`, so we need to write `e! & E.SomeFlag`.
	- The `no-unnecessary-type-assertion` lint rule ensures that these are only added where necessary.
	- Suspicious uses of `!` are marked with #18217.
* Adding `!!` to make expressions be `boolean` instead of `boolean | undefined`.
	- Also sometimes need to add `= false` to initialize the boolean.
	- Also change `condition && optional` to `condition ? optional : undefined` to avoid getting a `Optional | undefined | false`.
* Change `Debug.fail()` to `return Debug.fail()` to tell the compiler that we definitely return from a function

### CODE REVIEW

**Comment by sandersn** [comment]
* Looks like the public API needs to be updated, which is why Travis fails.
* How does this change compile time of the compiler? I think this is higher value than the all-unions change, but I would still hope it's not too much slower than what we have now.

**Comment by weswigham** [comment]
@sandersn On my workstation `jake clean tests` goes from a wallclock time of 1m7s to 1m10s. Seems good to me.

**Comment by weswigham** [comment]
I'm on board with all the changes except the non-optional non-undefined parent-pointers. While internally we use them with reckless abandon (the checker only operates on fully bound trees, after all), people using our transforms APIs won't have parent pointers bound (and shouldn't use them) - and making our public API correct is important. The, like, 1000 extra assertions we'd need internally are _probably_ worth it.

Also `flags! | <some flags>` is a bit odd (I mean, the assertion is never going to go away)? Maybe we should probably be allowing `undefined` in binary arithmetic without an error?

**Comment by andy-ms** [comment]
API tests pass when I run them locally on either windows or linux -- don't know why travis is failing.

**Comment by weswigham** [comment]
The travis failure repros locally for me (looks like `tests\baselines\reference\api\typescript.d.ts` has more `undefined`'s than it needs). Do you, perhaps, have an different, cached copy of the input or baselined `typescript.d.ts` knocking around locally that just won't flush?

**Comment by andy-ms** [comment]
Just needed to get Jake to run with `--strictNullChecks` too, thanks for the help.

**Comment by weswigham** [comment]
TIL jenkins will refuse to build when there are merge conflicts.

File: Jakefile.js [hunk 1]
```diff
@@ -307,7 +307,7 @@ function compileFile(outFile, sources, prereqs, prefixes, useBuiltCompiler, opts
         else {
             options += " --lib es5";
         }
-        options += " --noUnusedLocals --noUnusedParameters";
+        options += " --noUnusedLocals --noUnusedParameters --strictNullChecks";
```
Comment (mhegazy) [inline Jakefile.js]: why not just `--strict`?

Comment (mhegazy) [reply Jakefile.js]: and same in gulpfile

Comment (mhegazy) [reply Jakefile.js]: and same in tsconfig.json

File: scripts/buildProtocol.ts [hunk 2]
```diff
@@ -178,7 +178,7 @@ function writeProtocolFile(outputFile: string, protocolTs: string, typeScriptSer
     ts.sys.writeFile(outputFile, protocolDts);
 
     if (diagnostics.length) {
-        const flattenedDiagnostics = diagnostics.map(d => `${ts.flattenDiagnosticMessageText(d.messageText, "\n")} at ${d.file.fileName} line ${d.start}`).join("\n");
+        const flattenedDiagnostics = diagnostics.map(d => `${ts.flattenDiagnosticMessageText(d.messageText, "\n")} at ${d.file!.fileName} line ${d.start!}`).join("\n");
```
Comment (mhegazy) [inline scripts/buildProtocol.ts]: should properly throw if file is not set.

File: scripts/tslint/rules/noInOperatorRule.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ export class Rule extends Lint.Rules.AbstractRule {
 function walk(ctx: Lint.WalkContext<void>): void {
     ts.forEachChild(ctx.sourceFile, recur);
     function recur(node: ts.Node): void {
-        if (node.kind === ts.SyntaxKind.InKeyword && node.parent.kind === ts.SyntaxKind.BinaryExpression) {
+        if (node.kind === ts.SyntaxKind.InKeyword && node.parent!.kind === ts.SyntaxKind.BinaryExpression) {
```
Comment (mhegazy) [inline scripts/tslint/rules/noInOperatorRule.ts]: parent is not optional, why is this required?

File: src/compiler/types.ts [hunk 4]
```diff
@@ -2139,12 +2139,12 @@ namespace ts {
 
     export interface JSDocNamespaceDeclaration extends ModuleDeclaration {
         name: Identifier;
-        body: JSDocNamespaceBody;
+        body: JSDocNamespaceBody | undefined;
```
Comment (mhegazy) [inline src/compiler/types.ts]: why not `body?: JSDocNamespaceBody` ?

File: src/compiler/types.ts [hunk 5]
```diff
@@ -1685,7 +1685,7 @@ namespace ts {
     export interface ElementAccessExpression extends MemberExpression {
         kind: SyntaxKind.ElementAccessExpression;
         expression: LeftHandSideExpression;
-        argumentExpression?: Expression;
+        argumentExpression: Expression;
```
Comment (ajafff) [inline src/compiler/types.ts]: `argumentExpression` can actually be undefined if you have a syntax error like `foo[]`

File: src/compiler/program.ts [hunk 6]
```diff
@@ -526,10 +525,10 @@ namespace ts {
 
         performance.mark("beforeProgram");
 
-        host = host || createCompilerHost(options);
+        const host = compilerHost || createCompilerHost(options);
```
Comment (ajafff) [inline src/compiler/program.ts]: why not simply parameter defaults?

File: src/compiler/sys.ts [hunk 7]
```diff
@@ -456,7 +456,7 @@ namespace ts {
         getMemoryUsage?(): number;
         exit(exitCode?: number): void;
         realpath?(path: string): string;
-        /*@internal*/ getEnvironmentVariable(name: string): string | undefined;
```
Comment (andy-ms) [inline src/compiler/sys.ts]: Isn't this going to be confusing if external clients use `undefined` to represent a missing variable but we use `""`?

File: src/compiler/types.ts (hunk 5 above)
Comment (mhegazy) [reply src/compiler/types.ts]: we should create a missing node instead here. almost all the code afterwords does not handel it being undefined.

File: src/compiler/core.ts [hunk 8]
```diff
@@ -59,14 +59,14 @@ namespace ts {
...
     }
 
-    export function createMapFromTemplate<T>(template?: MapLike<T>): Map<T> {
+    export function createMapFromTemplate<T>(template: MapLike<T>): Map<T> {
         const map: Map<T> = new MapCtr<T>();
 
         // Copies keys/values from template. Note that for..in will not throw if
         // template is undefined, and instead will just exit the loop.
-        for (const key in template!) {
+        for (const key in template) {
```
Comment (andy-ms) [inline src/compiler/core.ts]: `for-in` actually works on `undefined` inputs, so this will work if `template` is undefined. See #19461

File: src/server/editorServices.ts [hunk 9]
```diff
@@ -1333,7 +1333,7 @@ namespace ts.server {
                 configHasFilesProperty: parsedCommandLine.raw.files !== undefined,
                 configHasIncludeProperty: parsedCommandLine.raw.include !== undefined,
                 configHasExcludeProperty: parsedCommandLine.raw.exclude !== undefined,
-                wildcardDirectories: createMapFromTemplate(parsedCommandLine.wildcardDirectories),
+                wildcardDirectories: createMapFromTemplate(parsedCommandLine.wildcardDirectories!), // TODO: GH#18217
```
Comment (andy-ms) [inline src/server/editorServices.ts]: That function should work for undefined inputs.

Comment (mhegazy) [reply src/server/editorServices.ts]: We do not allow undefined in `for (... in x)`. we should.

File: scripts/tslint/rules/noInOperatorRule.ts (hunk 3 above)
Comment (andy-ms) [reply scripts/tslint/rules/noInOperatorRule.ts]: This uses our published `typescript.d.ts`, not the one in this PR -- so if we merge this and the newest published version has `parent` non-optional, we'll have to change it then.

File: src/compiler/types.ts [hunk 10]
```diff
@@ -3371,8 +3371,8 @@ namespace ts {
     export interface Symbol {
         flags: SymbolFlags;                     // Symbol flags
         escapedName: __String;                  // Name of symbol
-        declarations?: Declaration[];           // Declarations associated with this symbol
-        valueDeclaration?: Declaration;         // First value declaration of the symbol
+        declarations: Declaration[];           // Declarations associated with this symbol
+        valueDeclaration: Declaration;         // First value declaration of the symbol
```
Comment (ajafff) [inline src/compiler/types.ts]: IIRC `valueDeclaration` is undefined for interfaces (and types?), is that still the case?

Comment (weswigham) [reply src/compiler/types.ts]: Yes.

Comment (andy-ms) [reply src/compiler/types.ts]: Yes, the change was to remove a large number of assertions even if it meant having a less accurate type declaration.

File: src/compiler/types.ts [hunk 11]
```diff
@@ -3371,8 +3371,8 @@ namespace ts {
     export interface Symbol {
         flags: SymbolFlags;                     // Symbol flags
         escapedName: __String;                  // Name of symbol
-        declarations: Declaration[];           // Declarations associated with this symbol
-        valueDeclaration: Declaration;         // First value declaration of the symbol
+        declarations: Declaration[];            // Declarations associated with this symbol
+        valueDeclaration: Declaration;          // First value declaration of the symbol
```
Comment (ajafff) [inline src/compiler/types.ts]: I still don't think this is the correct thing to do @andy-ms 

> Yes, the change was to remove a large number of assertions even if it meant having a less accurate type declaration.

If that property was internal, I wouldn't have a problem with that. But this makes the API really dangerous to use. Using `--strictNullChecks` in the compiler's code base should actually increase type safety for API users instead of decreasing it.

With the same reasoning you could make `Node.parent` non-optional to safe a lot of assertions, because it will be defined *almost* all the time.

Comment (andy-ms) [reply src/compiler/types.ts]: It's possible that we could change the output `.d.ts` files in a postprocessing step. CC @weswigham

File: Jakefile.js (hunk 1 above)
Comment (andy-ms) [reply Jakefile.js]: `--strictPropertyInitialization` gives a ton of errors, as does `--strictFunctionTypes`.
